### PR TITLE
[CS2113T-F12-3] Teaching Assistant's Assistant

### PR DESCRIPTION
**Pull request by hozhenhong99**
Teaching Assistant's Assistant helps teaching assistants manage details of their students. It is optimised for CLI users so that managing information can be done faster by typing in commands.

### CODE REVIEW

File: docs/DeveloperGuide.md [hunk 1]
```diff
@@ -1,38 +1,228 @@
...
+  * On run: Loads persistent data from `Storage`, receives user input from `UI`, and uses `Parser` to parse the user input.
+* `Parser`
+  * Handles input parsing and determines which command to run.
+* `Command`
+  * Defines how a command is to be executed.
+* `ModuleList`
+  * Contains a list of `Module` objects currently being managed by the app.
+* `Util`
+  * Contains useful methods (e.g. Check if a string is integer/double, convert string to integer/double).
+* `Storage`
```
Comment (Rakesh12000) [inline docs/DeveloperGuide.md:54]: Youd could add UML diagrams for the Storage function. This will enable your users to understand how data is written and stored by TAA as well as how it is accessed an read as inputs.

File: docs/DeveloperGuide.md [hunk 2]
```diff
@@ -1,38 +1,228 @@
...
+* Step 1 - The user executes `add_student c/CS2113T i/a0217978j n/jonny` to add a student. The `add_student` command 
+calls the `AddStudentCommand#execute` method. Within `AddStudentCommand#execute`, `ModuleList#getModule("CS2113T")` is 
+called to ensure that there is an existing module with code `CS2113T`.
+* Step 2 - If an existing module with code `CS2113T` is found, a new `Student` object with id and name set to 
+`a0217978j` and `jonny` respectively. Then, `StudentList#addModule` is called to add the newly created `Student` 
+object into the `students` ArrayList within `StudentList`.
+
+<br>
+
+### Add Assessment
```
Comment (Rakesh12000) [inline docs/DeveloperGuide.md]: The description for the implementation of Assessment tends to be a bit wordy and hard to follow. To make it easier for your users to understand this command function you could add an UML sequence diagram like the other commands.

File: docs/DeveloperGuide.md [hunk 3]
```diff
@@ -1,38 +1,228 @@
...
+* Step 4 - Within `AssessmentList#addAssessment`, the name of the newly created `Assessment` object is checked to ensure
+  there is no existing assessment with name `midterms`. At the same time, the weightage of the newly created
+  `Assessment` object is also checked to ensure that the total weightage of the assessments in the `cs2113t` module
+  will not exceed 100 with the addition of the weightage of the newly created `Assessment` object.
+* Step 5 - If the name and weightage of the newly created `Assessment` object are valid, the newly created `Assessment`
+  object is added into the `assessments` ArrayList within `AssessmentList`.
+
+<br>
+
+### Set Marks
```
Comment (Rakesh12000) [inline docs/DeveloperGuide.md]: You could add an example scenario like the other commands so that it is easier for your users to understand how this command works.

File: docs/DeveloperGuide.md [hunk 4]
```diff
@@ -1,38 +1,228 @@
...
+* `Taa`
+  * On creation: Initializes the `UI` and `Storage` components.
+  * On run: Loads persistent data from `Storage`, receives user input from `UI`, and uses `Parser` to parse the user input.
+* `Parser`
+  * Handles input parsing and determines which command to run.
+* `Command`
+  * Defines how a command is to be executed.
+* `ModuleList`
+  * Contains a list of `Module` objects currently being managed by the app.
+* `Util`
```
Comment (Rakesh12000) [inline docs/DeveloperGuide.md:52]: You could further explain in-depth how the util component works. This will help your users get an idea of what util can be used for when they work on your code.

File: docs/DeveloperGuide.md [hunk 5]
```diff
@@ -1,38 +1,228 @@
...
+`getModuleWithCode(moduleCode:String)`.
+6. Using `module:Module`, we are able to get the `studentList:StudentList` and `assessmentList:AssessmentList`.
+7. The `student:Student` and `assessment:Assessment` are then retrieved from their lists using 
+`getStudentAt(studentIndex:Integer)` and `getAssessment(assessmentName:String)` respectively.
+8. Finally, `setMarks(assessmentName:String, marks:Double)` is called to set the marks of the assessment in the 
+student's `results` HashMap.
+9. A message will then be printed out to indicate to the user that the marks have been set successfully.
+
+
+### Set Attendance
```
Comment (Rakesh12000) [inline docs/DeveloperGuide.md:268]: You could use an UML sequence diagram here as well for easier reference of how this command function works for your users.

**Review by Rakesh12000 — COMMENTED** [review]
DG Review

File: docs/DeveloperGuide.md [hunk 6]
```diff
@@ -1,38 +1,228 @@
...
+* `StudentList#getStudents()` - Returns an ArrayList containing all the students.
+* `StudentList#getStudentAt(index:int)` - Returns a student with the particular index.
+* `StudentList#isValidIndex(index:int)` - Checks if an index is valid w.r.t the `students` ArrayList.
+* `StudentList#addStudent(student:Student)` - Adds a student to the `students` ArrayList.
+* `StudentList#deleteStudent(index:int)` - Deletes the `student` object at the specified `index` within the `students` 
+ArrayList.
+* `StudentList#findStudent(keyword:String)` - Returns an ArrayList of students containing the keyword
+
+The sequence diagram shown below illustrates how the `add_student` command works:
+![AddStudentSequenceDiagram](diagrams/AddStudentSequenceDiagram.png)
```
Comment (BryanElmer) [inline docs/DeveloperGuide.md]: You could maybe add numbering in the diagram so that you can link the steps to the diagram to make it more clear.

File: docs/DeveloperGuide.md [hunk 7]
```diff
@@ -1,38 +1,228 @@
...
+
+**Interaction between components**
+![ArchitectureSequenceDiagram](diagrams/ArchitectureSequenceDiagram.png)
+
+The _Architecture Sequence Diagram_ shown above shows how the components usually interact with each other. 
+
+## Implementation
+### Add Module
+The sequence diagram shown below illustrates how the `add_module` command works:
+![AddModuleSequenceDiagram](diagrams/AddModuleSequenceDiagram.png)
```
Comment (BryanElmer) [inline docs/DeveloperGuide.md]: Nice description for the steps, but maybe you can add numbering in the diagram and link it to the steps.

File: docs/DeveloperGuide.md [hunk 8]
```diff
@@ -1,38 +1,228 @@
...
+  `Assessment` object is also checked to ensure that the total weightage of the assessments in the `cs2113t` module
+  will not exceed 100 with the addition of the weightage of the newly created `Assessment` object.
+* Step 5 - If the name and weightage of the newly created `Assessment` object are valid, the newly created `Assessment`
+  object is added into the `assessments` ArrayList within `AssessmentList`.
+
+<br>
+
+### Set Marks
+The sequence diagram shown below illustrates how the `set_mark` command works:
+![SetMarkSequenceDiagram](diagrams/SetMarkSequenceDiagram.png)
```
Comment (BryanElmer) [inline docs/DeveloperGuide.md]: You could maybe also add numbering for the diagram so that readers can easily refer to the diagram from the steps.

File: docs/DeveloperGuide.md [hunk 9]
```diff
@@ -1,38 +1,228 @@
...
+* `AttendanceList#getAttendance(lessonNumber:String)` - Returns an attendance with a particular lesson number.
+* `AttendanceList#isValidIndex(index:int)` - Checks if an index is valid w.r.t the `attendances` ArrayList.
+* `AttendanceList#addAttendance(attendance:Attendance)` - Adds an attendance to the `attendances` ArrayList.
+* `AttendnaceList#getAttendnaceIndex(lessonNumInput:String)` - Returns the attendance index in the `attendances` 
+ArrayList.
+* `AttendanceList#deleteAttendance(lessonNumInput:String)` - Deletes an attendance with a particular lesson number.
+* `AttendanceList#sortAttendances` - Sorts the attendance in the `attendances` ArrayList in ascending order based on 
+lesson number.
+
+Below is an example scenario of how the set attendance feature behaves at each step:
```
Comment (BryanElmer) [inline docs/DeveloperGuide.md:327]: The steps here look a bit chunky, maybe you could split them up into more steps which could help readers digest it better.

File: docs/DeveloperGuide.md [hunk 10]
```diff
@@ -1,38 +1,228 @@
 # Developer Guide
```
Comment (conradwee) [inline docs/DeveloperGuide.md:1]: great work!

File: docs/DeveloperGuide.md [hunk 11]
```diff
@@ -1,38 +1,228 @@
...
 
 {list here sources of all reused/adapted ideas, code, documentation, and third-party libraries -- include links to the original source as well}
 
-## Design & implementation
-
-{Describe the design and implementation of the product. Use UML diagrams and short code snippets where applicable.}
-
+## Design
+### Architecture
+![ArchitectureDiagram](diagrams/ArchitectureDiagram.png)
```
Comment (conradwee) [inline docs/DeveloperGuide.md:34]: would it be better to use UML here?

File: docs/DeveloperGuide.md (hunk 2 above)
Comment (conradwee) [inline docs/DeveloperGuide.md]: Neatness/correctness:
looks messy
not well-formatted
hard to read/understand

File: docs/DeveloperGuide.md (hunk 3 above)
Comment (conradwee) [inline docs/DeveloperGuide.md]: Use of examples:

Not enough or too many examples e.g., sample inputs/outputs

File: docs/DeveloperGuide.md [hunk 12]
```diff
@@ -1,38 +1,228 @@
...
+* `AttendanceList#isValidIndex(index:int)` - Checks if an index is valid w.r.t the `attendances` ArrayList.
+* `AttendanceList#addAttendance(attendance:Attendance)` - Adds an attendance to the `attendances` ArrayList.
+* `AttendnaceList#getAttendnaceIndex(lessonNumInput:String)` - Returns the attendance index in the `attendances` 
+ArrayList.
+* `AttendanceList#deleteAttendance(lessonNumInput:String)` - Deletes an attendance with a particular lesson number.
+* `AttendanceList#sortAttendances` - Sorts the attendance in the `attendances` ArrayList in ascending order based on 
+lesson number.
+
+Below is an example scenario of how the set attendance feature behaves at each step:
+* Step 1 - The user executes `set_attendance c/CS2113T s/1 l/1 p/1` to set an attendance to `Present` for student at
```
Comment (conradwee) [inline docs/DeveloperGuide.md]: Neatness/correctness:
looks messy
not well-formatted
hard to read/understand

perhaps some more spacing at a minimum

**Review by conradwee — COMMENTED** [review]
this looks like a great job, just make some minor changes and alls good

File: docs/DeveloperGuide.md [hunk 13]
```diff
@@ -1,38 +1,228 @@
...
+2. A user with above average typing speed for regular English text (i.e. not code, not system admin commands) should 
+be able to accomplish most of the tasks faster using commands than using the mouse.
 
 ## Glossary
-
-* *glossary item* - Definition
-
-## Instructions for manual testing
-
-{Give instructions on how to do a manual product testing e.g., how to load sample data to be used for testing}
```
Comment (Nirmaleshwar) [inline docs/DeveloperGuide.md:38]: I'm not sure if you're going you add this later but, perhaps adding some simple manual testing directions can guide the person to test functions you've described so far.

File: docs/DeveloperGuide.md [hunk 14]
```diff
@@ -1,38 +1,228 @@
...
+`AttendanceList` implements the following methods:
+* `AttendanceList#getSize()` - Returns the no. of attendance currently in the list.
+* `AttendanceList#getAttendances()` - Returns an ArrayList containing all the attendances.
+* `AttendanceList#getAttendance(lessonNumber:String)` - Returns an attendance with a particular lesson number.
+* `AttendanceList#isValidIndex(index:int)` - Checks if an index is valid w.r.t the `attendances` ArrayList.
+* `AttendanceList#addAttendance(attendance:Attendance)` - Adds an attendance to the `attendances` ArrayList.
+* `AttendnaceList#getAttendnaceIndex(lessonNumInput:String)` - Returns the attendance index in the `attendances` 
+ArrayList.
+* `AttendanceList#deleteAttendance(lessonNumInput:String)` - Deletes an attendance with a particular lesson number.
+* `AttendanceList#sortAttendances` - Sorts the attendance in the `attendances` ArrayList in ascending order based on
```
Comment (Nirmaleshwar) [inline docs/DeveloperGuide.md]: Maybe can consider putting these on a table as they are a bit too hard to read.

**Review by Nirmaleshwar — COMMENTED** [review]
Overall a very good job, just a few minor changes but since it's a work in progress, it's well on track.

File: docs/DeveloperGuide.md [hunk 15]
```diff
@@ -1,38 +1,228 @@
...
+![ArchitectureDiagram](diagrams/ArchitectureDiagram.png)
+
+The _Architecture Diagram_ shown above illustrates the high-level design of the Application.
+
+**Overview of components**
+* `Main`
+  * On app launch: Creates and runs an instance of `Taa`.
+* `UI`
+  * Handles UI operations.
+* `Taa`
```
Comment (peng-217) [inline docs/DeveloperGuide.md:41]: The sequence diagram of TAA has a problem:
 
image

The activation bar of Taa should be closed

File: docs/DeveloperGuide.md [hunk 16]
```diff
@@ -1,38 +1,228 @@
...
+  * On creation: Initializes the `UI` and `Storage` components.
+  * On run: Loads persistent data from `Storage`, receives user input from `UI`, and uses `Parser` to parse the user input.
+* `Parser`
+  * Handles input parsing and determines which command to run.
+* `Command`
```
Comment (peng-217) [inline docs/DeveloperGuide.md:48]: Would it be better to add function notation?

image

File: docs/DeveloperGuide.md [hunk 17]
```diff
@@ -1,38 +1,228 @@
...
+  * Contains a list of `Module` objects currently being managed by the app.
+* `Util`
+  * Contains useful methods (e.g. Check if a string is integer/double, convert string to integer/double).
+* `Storage`
+  * Handles data storage operations (e.g. Reading from and writing to data file).
+
+<br>
+
+**Interaction between components**
+![ArchitectureSequenceDiagram](diagrams/ArchitectureSequenceDiagram.png)
```
Comment (peng-217) [inline docs/DeveloperGuide.md]: It would be better to add the notation of the grey file icon
image

File: docs/DeveloperGuide.md (hunk 16 above)
Comment (peng-217) [reply docs/DeveloperGuide.md:48]: Better to add function notation
image

File: docs/DeveloperGuide.md [hunk 18]
```diff
@@ -1,38 +1,231 @@
...
+* `Parser`
+  * Handles input parsing and determines which command to run.
+* `Command`
+  * Defines how a command is to be executed.
+* `ModuleList`
+  * Contains a list of `Module` objects currently being managed by the app.
+* `Util`
+  * Contains useful methods (e.g. Check if a string is integer/double, convert string to integer/double).
+* `Storage`
+  * Handles data storage operations (e.g. Reading from and writing to data file).
```
Comment (tttyyzzz) [inline docs/DeveloperGuide.md:55]: image
Perhaps adding the new ModuleList()

Comment (tttyyzzz) [reply docs/DeveloperGuide.md:55]: image
Perhaps adding new Command() upon instantiation

File: docs/DeveloperGuide.md [hunk 19]
```diff
@@ -1,38 +1,231 @@
...
+* `AssessmentList#getSize()` - Returns the no. of assessments currently in the list.
+* `AssessmentList#getAssessments()` - Returns an ArrayList containing all the assessments.
+* `AssessmentList#getAssessment(assessmentName:String)` - Returns an assessment with a particular name specified by the
+  user.
+* `AssessmentList#addAssessment(assessment:Assessment)` - Adds an assessment to the `assessments` ArrayList.
+* `AssessmentList#deleteAssessment(assessmentName:String)` - Deletes an assessment with the name specified by the user
+  from the `assessments` ArrayList.
+
+The sequence diagram shown below illustrates how the `add_assessment` command works:
+![AddAssessmentSequenceDiagram](diagrams/AddAssessmentSequenceDiagram.png)
```
Comment (tttyyzzz) [inline docs/DeveloperGuide.md]: image
Perhaps adding the 'X' below the addModuleCommand

File: docs/DeveloperGuide.md [hunk 20]
```diff
@@ -1,38 +1,231 @@
...
+
+**Interaction between components**
+![ArchitectureSequenceDiagram](diagrams/ArchitectureSequenceDiagram.png)
+
+The _Architecture Sequence Diagram_ shown above shows how the components usually interact with each other. 
+
+## Implementation
+### Add Module
+The sequence diagram shown below illustrates how the `add_module` command works:
+![AddModuleSequenceDiagram](diagrams/AddModuleSequenceDiagram.png)
```
Comment (tttyyzzz) [inline docs/DeveloperGuide.md]: image
Perhaps deleting the redundant class.

File: docs/DeveloperGuide.md (hunk 6 above)
Comment (tttyyzzz) [reply docs/DeveloperGuide.md]: Perhaps you could explain step by step how the add_student command is being parsed and saved.

Comment (tttyyzzz) [reply docs/DeveloperGuide.md]: You could perhaps split up the uml diagram to include sub parts of the e.g. focuses on the logic only so that the uml diagram could be bigger and clearer. ref: https://se-education.org/addressbook-level3/DeveloperGuide.html

**Review by tttyyzzz — COMMENTED** [review]
Overall good uml diagrams with detailed labels. Could use numbering in uml diagrams if it helps in explanation. Perhaps standardise the steps underneath the uml diagrams. Good Job!
